### ISSUE TITLE
Fan Speeds - Inconsistent

### ISSUE THREAD

**Issue by patrickoleary1977** [aarons22/homebridge-bond#55]
Hey there, awesome plugin.  

Lights on and off work perfect for me.

Fan speed on the other hand is very inconsistent.  I have a 3 speed fan that works perfect in the Bond app. If I tell Siri, 100%...it will set it to low.  From there, if I say 33%, it will set to medium, etc.  If I shut off and then tell Siri 33%, it will set to 66%.  I can't seem to find a rhyme or reason for it.  Happy to help debug or grab any information -- though I may need help getting it.  I'm not exactly experienced with HomeBridge.

In my Bond app, I have speed 1, 2, 3.  Selecting one of those takes the fan directly to that speed.  

Thanks.

**Comment by fancypants66** [comment]
Agreed awesome job, love this plugin as well!

However, I've also got a 3-speed fan on v2.5.2, and while the light works as expected, the fan-speeds not responding at all from the Home app. I downgraded back to homebridge-bond@2.5.0 and fan-speeds are working as expected.

**Comment by shbui** [comment]
The issue is when retrieving fan speeds from Device, bond API returns an array of fan speeds in random order. The fix is to sort fan speeds before using them. Here are steps:

1. open index.ts
2. search for a line that contains 

const values = Device.fanSpeeds(device);

replace it with 
const values = [1,2,3]; //for 3-speed fan

or
const values = [1,2,3,4]; //for 4-speed fan

3. Restart homebridge.

In general, output of Device.fanSpeeds(device) should be sorted before assigned to values variable.

**Comment by fancypants66** [comment]
Thanks for the info @shbui - I'll give it a go!

**Comment by chrismerck** [comment]
It sounds like there's a dependence somewhere on the Bond `devices/<id>/commands` API. I would not recommend that, as this API is really just for the Bond Home app and not intended for integration use.

Instead, one should use `devices/<id>/properties.max_speed` to determine the number of speeds that the fan has.

Also note that, if you avoid dependence on commands, then the same integration should work for  Smart by Bond fans, too.

**Comment by jimboslicethat** [comment]
I'm experiencing the same thing on the latest version `2.5.2`. I have a 3 speed fan My 100% speed is now my 33%  speed, my 33% speed is now my 100% speed, and my 66% speed stayed the same.

**Comment by chrismerck** [comment]
Hey James (@jimboslicethat) FWIW, v2.5.2 is very old "MVP" firmware. I cannot say confidently that we've fixed a bug related to this specific issue since v2.5.2, but we are up to v2.10.8 and have a tonne of improvements! Perhaps you are on the old Bond Bridge (V1) app rather than the new Bond Home (V2) app?

**Comment by shbui** [comment]
I saw this issue on my 3-fan speed with Bond v2.9.5. When using Postman to query fan speeds from Bond, I saw Bond returning an array of fan speeds with values in random order. It should be [1,2,3]. The calculations in homebridge-bond index.ts are based on an order of fan speed, therefore, my fan speed is consistent. To fix this, values returned by calling Device.fanSpeeds(device) in index.ts should be sorted before they are used in calculations. I have tested on my 3-fan speed and it works fine after this fix. I have not checked with Bond v.2.10.8 yet.

**Comment by jimboslicethat** [comment]
@chrismerck I started noticing this issue when I upgraded my raspberry pi to have version 2.5.2 of homebridge-bond plugin. My Bond Home is completely up to date with firmware version 2.10.8.

**Comment by patrickoleary1977** [comment]
> The issue is when retrieving fan speeds from Device, bond API returns an array of fan speeds in random order. The fix is to sort fan speeds before using them. Here are steps:
> 
> 1. open index.ts
> 2. search for a line that contains
> 
> const values = Device.fanSpeeds(device);
> 
> replace it with
> const values = [1,2,3]; //for 3-speed fan
> 
> or
> const values = [1,2,3,4]; //for 4-speed fan
> 
> 1. Restart homebridge.
> 
> In general, output of Device.fanSpeeds(device) should be sorted before assigned to values variable.

That seems to have worked for me on the 33% and 100%.  However, it still won't go to medium i.e. 67%.

**Comment by aarons22** [comment]
@shbui thanks for the investigation here! I'll get out a fix to sort the values from the API. In the future feel free to throw up a pull request and tag me for review as I'm not always checking in on this repo 😄

**Comment by aarons22** [comment]
@chrismerck the dependency on the commands API was discussed in [this issue](https://github.com/aarons22/homebridge-bond/issues/13) and in the [bond forum](https://forum.bondhome.io/t/smart-by-bond-api-command-not-working/907/13?u=aarons22).

TLDR: using `max_speed` is great, but it assumes the user added all speeds in the sequence.

Also, I believe the folks in this thread were referring to v2.5.2 of this plugin 😅

**Comment by aarons22** [comment]
@shbui Well that was easy. I just pushed a fix in [2.5.3](https://github.com/aarons22/homebridge-bond/releases/tag/2.5.3). Can you try it out and see if it works as you would expect?

**Comment by fancypants66** [comment]
I think I may be experiencing a different issue. Running either 2.5.2 or 2.5.3 of homebridge-bond, fan lights continue to work as expected - but changing fan speeds in either Home app or Siri and nothing happens?? The Debug log is completely blank as if no fan speed command was ever sent. Bond is on firmware v2.10.8. Downgrading back to homebridge-bond@2.5.0 and fan speeds work as expected. Is there a possible issue with the fan speed commands in the Bond app? I'll try re-adding the fan there...

**Comment by patrickoleary1977** [comment]
I updated.  That didn’t work for me.  In Home, if I set to 100% the fan comes on high, if i set to 67% the fan goes to low, if i set to 33% the fan goes to medium.

**Comment by fancypants66** [comment]
Ok - I cleared out my accessories/persist cache in homebridge and 2.5.3 fan speeds are now working correctly for me. 3-speed fan. Debug commands for fan speeds are now logging for me.

**Comment by patrickoleary1977** [comment]
How can I do the same? Sorry, homebridge newbie.

**Comment by fancypants66** [comment]
@patrickoleary1977 depends your homebridge setup - but on a raspberry pi, I've got it in these locations: /var/lib/homebridge/accessories/ and /var/lib/homebridge/accessories/persist/. It's a good idea to make a copy before clearing it out just incase something goes awfully wrong ;)
